File: hospital-ai/frontend/src/types/ocr.ts
/** Intake Stage 3 — Document Processing / OCR types. */

import type { ProcessingJob } from '@/types/registration';
import type { UnifiedMedicalHistory, TimelineEvent } from '@/types/medicalHistory';

export interface OCRResult {
  id: string;
  processing_job_id: string;
  patient_id: string;
  document_id?: string | null;
  provider: string;
  raw_text: string;
  clean_text?: string | null;
  tables_json: unknown[];
  images_json: unknown[];
  detected_language?: string | null;
  page_count: number;
  processing_time_ms?: number | null;
  confidence?: number | null;
  provenance_json?: Record<string, unknown>;
  extraction_method?: string | null;
  processing_method?: string | null;
  document_type?: string | null;
  ocr_provider?: string | null;
  library_used?: string | null;
  fallback_used?: boolean | null;
  document_status?: string | null;
  status?: string | null;
  character_count?: number | null;
  word_count?: number | null;
  processing_logs?: unknown[] | null;
  error_message?: string | null;
  created_at: string;
}

export interface OCRStartResult {
  job_id: string;
  patient_id: string;
  status: string;
  current_stage: string;
  next_stage: string;
  provider: string;
  confidence: number;
  page_count: number;
  processing_time_ms: number;
  detected_language: string;
  extraction_method?: string;
  processing_method?: string;
  document_type?: string;
  ocr_provider?: string | null;
  library_used?: string;
  fallback_used?: boolean;
  document_status?: string;
  character_count?: number;
  word_count?: number;
  warnings: string[];
  ocr_result: OCRResult;
  processing_job: ProcessingJob;
  patient_context_version: number;
}

export interface OCRStatus {
  job_id: string;
  status: string;
  current_stage: string;
  next_stage: string;
  progress_pct: number;
  provider?: string | null;
  confidence?: number | null;
  processing_time_ms?: number | null;
  error_message?: string | null;
  ocr_completed: boolean;
  processing_job: ProcessingJob;
}

export interface PatientContextView {
  patient_id: string;
  status: string;
  current_summary?: string | null;
  ocr_completed: boolean;
  ocr_provider?: string | null;
  ocr_confidence?: number | null;
  ocr_timestamp?: string | null;
  context_version: number;
  patient_context_json: Record<string, unknown>;
  metadata: Record<string, unknown>;
}

export type IntakeStageId =
  | 'registration'
  | 'history'
  | 'context'
  | 'ocr'
  | 'ner'
  | 'risk'
  | 'knowledge';

export interface IntakeStageDef {
  id: IntakeStageId;
  label: string;
  description?: string;
  stageKey?: string;
}

export const INTAKE_STAGES: IntakeStageDef[] = [
  {
    id: 'registration',
    label: 'Patient Registration',
    stageKey: 'Patient Registration',
    description: 'Link patient, appointment, doctor, and upload the report.',
  },
  {
    id: 'history',
    label: 'Medical History Extraction',
    stageKey: 'Medical History Extraction',
    description: 'Gather prior visits, diagnoses, and medications from hospital records.',
  },
  {
    id: 'context',
    label: 'Patient Context Builder',
    description: 'Build a readable clinical summary for downstream AI stages.',
  },
  {
    id: 'ocr',
    label: 'Document Processing',
    stageKey: 'OCR',
    description: 'Convert uploaded documents into readable text.',
  },
  {
    id: 'ner',
    label: 'Medical Entity Recognition',
    stageKey: 'Medical Entity Recognition',
    description: 'Identify clinical entities from the extracted text.',
  },
  {
    id: 'risk',
    label: 'Patient Risk Profiling',
    description: 'Estimate clinical risk from structured findings.',
  },
  {
    id: 'knowledge',
    label: 'Patient Knowledge Graph',
    description: 'Connect symptoms, conditions, and treatments for AI agents.',
  },
];

export type { UnifiedMedicalHistory, TimelineEvent };

File: hospital-ai/backend/app/schemas/ocr.py
"""Pydantic schemas for Intake Stage 3 — Document Processing / OCR."""

from datetime import datetime
from typing import Any, Dict, List, Optional
from uuid import UUID

from pydantic import BaseModel, Field

from app.schemas.registration import ProcessingJobOut


class OCRStartRequest(BaseModel):
    job_id: UUID


class OCRResultOut(BaseModel):
    id: UUID
    processing_job_id: UUID
    patient_id: UUID
    document_id: Optional[UUID] = None
    provider: str
    raw_text: str
    clean_text: Optional[str] = None
    tables_json: List[Any] = Field(default_factory=list)
    images_json: List[Any] = Field(default_factory=list)
    detected_language: Optional[str] = None
    page_count: int = 1
    processing_time_ms: Optional[int] = None
    confidence: Optional[float] = None
    provenance_json: Dict[str, Any] = Field(default_factory=dict)
    extraction_method: Optional[str] = None
    processing_method: Optional[str] = None
    document_type: Optional[str] = None
    ocr_provider: Optional[str] = None
    library_used: Optional[str] = None
    fallback_used: Optional[bool] = None
    document_status: Optional[str] = None
    status: Optional[str] = None
    character_count: Optional[int] = None
    word_count: Optional[int] = None
    processing_logs: Optional[List[Any]] = None
    error_message: Optional[str] = None
    created_at: datetime


class OCRStartResponse(BaseModel):
    job_id: UUID
    patient_id: UUID
    status: str
    current_stage: str
    next_stage: str = "Medical Entity Recognition"
    provider: str
    confidence: float
    page_count: int
    processing_time_ms: int
    detected_language: str
    extraction_method: str = "embedded_text"
    processing_method: str = "embedded_text"
    document_type: str = "digital_pdf"
    ocr_provider: Optional[str] = None
    library_used: str = ""
    fallback_used: bool = False
    document_status: str = "Completed"
    character_count: int = 0
    word_count: int = 0
    warnings: List[str] = Field(default_factory=list)
    ocr_result: OCRResultOut
    processing_job: ProcessingJobOut
    patient_context_version: int = 1


class OCRStatusResponse(BaseModel):
    job_id: UUID
    status: str
    current_stage: str
    next_stage: str
    progress_pct: int
    provider: Optional[str] = None
    confidence: Optional[float] = None
    processing_time_ms: Optional[int] = None
    error_message: Optional[str] = None
    ocr_completed: bool = False
    processing_job: ProcessingJobOut


class OCRReportOut(BaseModel):
    """Staff-facing medical report view — cleaned text only."""

    report_id: str
    file_name: str
    document_type: str
    processing_method: Optional[str] = None
    extraction_method: Optional[str] = None
    ocr_provider: Optional[str] = None
    upload_date: Optional[datetime] = None
    language: Optional[str] = None
    pages: int = 1
    confidence: Optional[float] = None
    processing_time_ms: Optional[int] = None
    status: str
    character_count: Optional[int] = None
    word_count: Optional[int] = None
    text_preview: str
    full_text: str


class PatientContextResponse(BaseModel):
    patient_id: UUID
    status: str
    current_summary: Optional[str] = None
    ocr_completed: bool = False
    ocr_provider: Optional[str] = None
    ocr_confidence: Optional[float] = None
    ocr_timestamp: Optional[datetime] = None
    context_version: int = 1
    patient_context_json: Dict[str, Any] = Field(default_factory=dict)
    metadata: Dict[str, Any] = Field(default_factory=dict)
